### DISCUSSION TITLE
Multiple Celery projects on same Redis database

### DISCUSSION THREAD

**Discussion by ca-simone-chiorazzo** [celery/celery#7152] [Q&A] [answered]
Hi all,

I have a question regarding Celery. We have a software system in which there are several services that use Celery using the same Redis instance as a broker.

In addition to that, we're sharing the same DB of celery for each installation and we've different versions of Celery/kombu for each service (different versions of Celery 4.X/ Kombu 4.X). The segregation that we are using is based on the queue names.

I saw that Kombu uses the same data structures to track the unacked messages that are not segregated by queue/project and so I'm afraid that can cause collisions/unexpected behaviours of Celery.

Is this considered a safe approach? Does Celery support this kind of approach?

Any advice is really appreciated!

Thanks,
Simone

**Comment by thedrow** [comment]
The safe approach would be to use different databases with the same instance.
`redis://localhost/0` for the first project and `redis://localhost/1` for the second.

Feel free to contribute documentation regarding multiple virtual hosts/databases.

**Reply by ca-simone-chiorazzo** [reply]
What kind of side effect there could be using the same DB?

**Reply by ca-simone-chiorazzo** [reply]
@thedrow I'd be happy to contribute to the documentation but I'd like also to understand why this approach is not suggested/possible

**Reply by thedrow** [reply]
For example, the celery default queue is by default `celery` and different workers can fetch work from that queue despite the fact that they are in different projects if you're not careful.
There are many pitfalls we haven't explored.

I strongly recommend to pick the easy solution and segregate the worker clusters logically.

**Comment by cedric-brunet** [comment]
Could we have an update on this ?
Is it safe or should we use the broker_transport_options "global_keyprefix" instead ?

**Comment by ElliotNguyen68** [comment]
Hi @auvipy @Nusnus  can we have answer on this ?

**Comment by oliverhaas** [comment, marked as answer]
`global_keyprefix` is exactly for this case to my knowledge. I don't see an upside _not_ using it.

From my (limited) experience with the code of the kombu redis transport, I would expect that the queue bindings will not work as expected, and if one service accidentally tries to restore an unacked message from another service, you might just lose your tasks.

**Comment by ElliotNguyen68** [comment]
yeah we implemented this with global_keyprefix, and it worked well.
